<a href="https://colab.research.google.com/github/wcj365/python-stats-dataviz/blob/master/lectures/14_Hypothesis_Testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 14 - Hypothesis Testing

Updated - today

**Null Hypothesis:** Status Quo

**Alternative Hypothesis:** Challenging the status quo

### One-Sample t-Test
A one-sample t-test checks whether a sample mean differs from the population mean. 

### Two-Sample t-Test
A two-sample t-test investigates whether the means of two independent data samples differ from one another. In a two-sample test, the null hypothesis is that the means of both groups are the same.

In [ ]:
import math
import scipy.stats as stats
import numpy as np

In [ ]:
# Generate three random samples from normal distribution.
# rvs stands for "random variable sample"
# The inputs are (population mean, population standard deviation, sample size)
np.random.seed(seed=250)
sample_A = stats.norm.rvs(10000, 1000, 50)
np.random.seed(seed=300)
sample_B = stats.norm.rvs(1000, 100, 50)
np.random.seed(seed=400)
sample_C = stats.norm.rvs(1000, 100, 50)

print("Sample A:\n\n", sample_A, end="\n\n")
print("Sample B: \n\n", sample_B, end="\n\n")
print("Sample C: \n\n", sample_C)

Sample A:

 [ 9299.87839497  9121.61369651 11273.18096915  8468.84367625
 11363.56887993 10900.52447749  8201.91619748 10868.97991022
 10627.28352542 11256.68481088  8960.52538184 10014.74714654
 10957.11006294  8890.66391574 10592.49721674  9979.42334971
 10335.5610067   9078.59451229 10076.80139397  9719.81012784
 11400.93603221 10401.88288405 10529.75604803  8474.19881925
  8885.46552282  9780.77318685 10784.06708957 10452.01671463
  9059.3531639   9888.40592509 11603.25236528  9875.11545266
  9278.77528989 10542.76434024  8829.47607725 10590.605351
 10407.13267933  9813.30494293 10987.78635357  8718.78226226
 10309.66017241 10471.2002314  10556.43823264 10040.87867831
  9602.76128026  9827.43486475  8885.24668895  9455.88401749
 10301.37148104  9559.34162071]

Sample B: 

 [ 851.48296373  973.97952481  857.54346303  910.86010695 1076.11633545
  930.1242738   960.33503496 1124.37873612 1007.45184628  967.70519662
 1146.14716267 1124.18213857 1025.94194149 1107.16883683  986.97819762

In [ ]:
print("Sample mean of sample A =", round(sample_A.mean(),2))
print("Sample mean of sample B =", round(sample_B.mean(),2))
print("Sample mean of sample C =", round(sample_C.mean(),2))

Sample mean of sample A = 9986.05
Sample mean of sample B = 1010.66
Sample mean of sample C = 1020.26


### One Sample t-Test

H0: the mean of the population that sample A drawn from is 0

Ha: the mean of the population that sample A drawn from is not 0


In [ ]:
stats.ttest_1samp(sample_A, 0)

Ttest_1sampResult(statistic=81.49018333519955, pvalue=5.540905729236289e-54)

p-value is the probability of having the sameple under the hull hypothesis.
If the population mean is 0 (null hypothesis), then the chance for the sample to have a mean of 9986.05 is very very slim - almost impossible.

H0: the mean of population that sample A sampled from is 9500

Ha: the mean of population that sample A sampled from is not 9500

In [ ]:
stats.ttest_1samp(sample_A, 9500)

Ttest_1sampResult(statistic=3.966328724225254, pvalue=0.00023756348939425685)

p-value is the probability of having the sameple under the hull hypothesis. If the population mean is 9500 (null hypothesis), then the chance for the sample to have a mean of 9986.05 is very slim.

H0: the mean of population that sample A sampled from is 9950

Ha: the mean of population that sample A sampled from is not 9950

In [ ]:
stats.ttest_1samp(sample_A, 9950)

Ttest_1sampResult(statistic=0.2941461373896291, pvalue=0.769888706845073)

p-value is the probability of having the sameple under the hull hypothesis. If the population mean is 9950 (null hypothesis), then the chance for the sample to have a mean of 10185 is about 12.6%, not slim. So we are unable to reject the null hypothesis. So we would conclude that given the evidence (sample data) the population mean is likely to be 9950 (we are still not 100% certain).

H0: the mean of population that sample A sampled from is 9900

Ha: the mean of population that sample A sampled from is not 9900

In [ ]:
stats.ttest_1samp(sample_A, 9900)

Ttest_1sampResult(statistic=0.7021664248158096, pvalue=0.48589760215283406)

### Two Sample Test - sample A vs sample B

H0: The populations that sample A and B were sampled from have the same mean

Ha: The populations that sample A and B were sampled from have different means

In [ ]:
stats.ttest_ind(sample_A,sample_B)

Ttest_indResult(statistic=72.6611958544629, pvalue=4.788266360211021e-87)

With very small tiny p-value, we reject the null hypothesis and accept the alternative hypothesis.

### Two Sample Test - sample B vs C

H0: The populations that sample B and C were sampled from have the same mean

Ha: The populations that sample B and C were sampled from have different means

In [ ]:
stats.ttest_ind(sample_B,sample_C)

Ttest_indResult(statistic=-0.4796162818907282, pvalue=0.632568913027603)

With very large p-value, we are unable to reject the null hypothesis. So we accept the null hypothesis. We have enough evidence to believe sample B and C were drawn from populations with the same mean.

### Two Sample Test - sample D vs E 

Let's make the population means different (1000 vs 1005)

- H0: The populations that sample D and E were sampled from have the same mean
- Ha: The populations that sample D and E were sampled from have different means

In [ ]:
for i in range(10):

    #np.random.seed(seed=250)
    sample_D = stats.norm.rvs(1000, 100, 5000)
    #np.random.seed(seed=350)
    sample_E = stats.norm.rvs(1015, 100, 5000)

    #print(sample_D)
    #print(sample_E)

    #print("Sample mean of sample D =", round(sample_D.mean(),2))
    #print("Sample mean of sample E =", round(sample_E.mean(),2))

    print("Diff =", round(sample_D.mean() - sample_E.mean(),2), stats.ttest_ind(sample_D, sample_E))

Diff = -16.7 Ttest_indResult(statistic=-8.26434377859484, pvalue=1.5829445351178002e-16)
Diff = -17.4 Ttest_indResult(statistic=-8.783470584168104, pvalue=1.8450647582108685e-18)
Diff = -14.77 Ttest_indResult(statistic=-7.406100958211381, pvalue=1.4057162662104136e-13)
Diff = -17.65 Ttest_indResult(statistic=-8.815803408671174, pvalue=1.3862843934834754e-18)
Diff = -13.05 Ttest_indResult(statistic=-6.526739400240044, pvalue=7.047617504165845e-11)
Diff = -16.33 Ttest_indResult(statistic=-8.12220459932874, pvalue=5.118467754934014e-16)
Diff = -16.58 Ttest_indResult(statistic=-8.3824675701816, pvalue=5.881044740829616e-17)
Diff = -17.98 Ttest_indResult(statistic=-8.978060384011565, pvalue=3.252105346592345e-19)
Diff = -14.96 Ttest_indResult(statistic=-7.487488469175645, pvalue=7.613270256757389e-14)
Diff = -13.26 Ttest_indResult(statistic=-6.6115377345470465, pvalue=3.997691762708448e-11)


Still relatively large p-value, so unable to reject the null hypothesis. 
Even though the population means are different but the difference is not statistically significant.


### Two Sample Test - sample D vs E 

Let's make the population means somewhat more different (1000 vs 1010)

H0: The populations that sample D and E were sampled from have the same mean
Ha: The populations that sample D and E were sampled from have different means

In [ ]:
sample_D = stats.norm.rvs(1000, 100, 50)
sample_E = stats.norm.rvs(1015, 100, 50)

stats.ttest_ind(sample_D, sample_E)

Ttest_indResult(statistic=-0.23023251455235985, pvalue=0.818391006052364)

The p-value is smaller but still greater than 0.05 (the standard used in social science research). We will not reject the null hypothesis. We conclude the populations are not significantly different.

### Two Sample Test - sample D vs E 

Let's make the population means even more different (1000 vs 1030)

H0: The populations that sample D and E were sampled from have the same mean
Ha: The populations that sample D and E were sampled from have different means

In [ ]:
sample_D = stats.norm.rvs(1000, 100, 50)
sample_E = stats.norm.rvs(1030, 100, 50)

stats.ttest_ind(sample_D, sample_E)

Ttest_indResult(statistic=-1.4255466585372618, pvalue=0.15717665187640334)

Now, we have a p-value that is less than standard 0.05.
We can reject the null hypothesis and state that the means of the two populations are not the same.

### The End